utils/_context/containers.py

```diff
@@ -493,6 +500,19 @@ def __init__(
             healthcheck={"test": f"curl --fail --silent --show-error localhost:{self.port}", "retries": 60},
             ports={"7777/tcp": self.port, "7778/tcp": weblog._grpc_port},
             stdout_interface=interfaces.library_stdout,
+            depends_on=[
```

Comment:
`depends_on` must be a list of `Testedcontainer` instances, not class. Other wise, if you have several container that are instance of the same class, but with different dependancies, it'll break.

Reply:
The only way to do that would be to define this externally, so when objects have been instantiated. This pushes the dependency management to the user of the class instead of having it right on the class itself. Since these are instantiated all over the place, it would pollute a lot of the code. I also disagree that it needs to be instances, as if the dependencies are well-defined, it should be possible to add them to the classes. For example, I would expect every BuddyContainer to have the same dependencies, and if that is no longer true, then I would expect to have something like a NodeBuddyContainer. If a class describes a container, then its container should always have the same dependencies.

Reply:
It's introducing a hidden constraint, it must not be done without a good reason. 

Actually, It's not needed, as the Testcontainer.depends_on is a public property, it does not have to be existing and complete at container creation, we can definitly use that : 

```
    my_container = SomeConatainer()
    self.weblog. depends_on.append(my_container)
```

Reply:
Yes but then you'd have to know about all the containers added to the array all over the code, keep track of that, and make sure references are either passed properly, or keep looping through the array to get the instance. It's just overall messy.

On the other hand, by having it on the class, it's centralized so it's very clean and easy to understand, and it allows doing things like saying weblog depends on all buddies by putting a dependency on BuddyContainer instead of on individual containers.

Reply:
> Yes but then you'd have to know about all the containers added to the array all over the code, keep track of that, and make sure references are either passed properly, or keep looping through the array to get the instance, and keep repeating that logic everywhere that dependencies exist. It's just overall messy.

Which is also true right now, you can do `self.weblog.depends_on.append(SomeContainer)`. In theory, at the end of `DockerScenario.__init__()`, the dependency tree is done, and should not be updated. We can enforce that. But again, what you describe is true for both solutions.

> by having it on the class, it's centralized

It'll be evenly centralized, in `__init__` method. If it's done correctly, you don't even need to use `depends_on.append`, it only needs to create containers in the right order. See the code above as an example.

> it's very clean and easy to understand

I do not agree. The link instance <-> class is very very very weak, and one could easily add another container, or change a container class, and have a hard time to understand why the dependency tree is brokken


Before
```python
class EndToEndScenario(DockerScenario):
    def __init__(self):
        super().__init__() # DockerScenario create the proxy
        self.weblog_container = WeblogContainer()
        self.agent_container = AgentContainer()
```

After, the only thing to change is that if A needs B, then B must be created before A in order to avoid a call to `TestedContainer.depends_on.append`

```python
class EndToEndScenario(DockerScenario):
    def __init__(self):
        super().__init__() # DockerScenario create the proxy
        self.agent_container = AgentContainer(needs=[self.poxy_container])
        self.weblog_container = WeblogContainer(use_proxy=use_proxy, needs=[self.agent_container])
```

Reply:
Another way of telling it : if you want to use the paradigm where its classes that ship info, and not  instances, you must use it everywhere, or nowhere. But the middle road where we use them sometimes as instance, sometimes as class, is not sustainable.

Reply:
It's never as instance with my approach, it's always by class. Maybe my mistake was adding it as an instance property. Would it work for you if I move `depends_on` to the class instead? My reasoning is that a class represents a _container_ and not an _image_. So a _container_ would always have the same dependencies.

Reply:
> It's never as instance with my approach, it's always by class.

In the current code, it's always by instances. We won't mix boths approach. I also add that having a list of `depends_on` that always contains stuff that are not started will make the data structure harder to understand.

Reply:
> In the current code, it's always by instances. We won't mix boths approach.

Yes that's what I'm asking, would you be fine with me moving to _always_ by class and _never_ by instance?

Reply:
Reasoning is that the only way to do it by container without using their names would be to have explicit references to every container, for example with `self._kafka_container` and so on because there are interdependencies across multiple methods. While doable, I think the class approach is simpler.

Reply:
No I'm not : 

* using instances is the natural way of doing. 
* Using classes will requires more advanced python tricks to do it cleanly, for a code harder to understand and maintain.
* it'll requires way more efforts to move everything to classes, compared to use the [code above](https://github.com/DataDog/system-tests/pull/2732#discussion_r1678201838).

I see absolutely no reason for using classes.

Reply:
> it'll requires way more efforts to move everything to classes,

No, it's literally the same as the current code for this PR, but with static properties instead.

> compared to use the https://github.com/DataDog/system-tests/pull/2732#discussion_r1678201838.

That code doesn't work, and making it work will require way more code and trickeries that this PR currently has.

Reply:
I think I have an idea to make this work with an external map. Let me try something.

Reply:
Nevermind that didn't work, the only way I see to make this work with containers is to add properties for each containers on the scenario, and then add dependencies everywhere as the container objects are being instantiated.

Reply:
I moved the whole logic to the caller so that it now works on the instances directly. This also removed the need to generate a dependency map as dependencies are now added directly by their dependents. Let me know if that's more in line with what you had in mind!

Reply:
> I moved the whole logic to the caller so that it now works on the instances directly.

Yes, it's  exactly what I meant, it's going into the right direction, and no need for any trickery.

> This also removed the need to generate a dependency map as dependencies are now added directly by their dependents.

Yes, one complexity removed.

Now, we need a very last thing to have a very clear and self explanatory data structure : first add back a `depends_on` property in `TestedContainer` then, every time you have 

```python
self._container_dependencies[container_a].append(container_b)
```

Your replace it by : 

```python
container_a.depends_on.append(container_b)
```

(On this example, `container_b` needs to be started before `container_a`)

And then, `_container_dependencies` can be removed.

We can also add it in the `TestedContainer` constructor, but using it would require a little bit more of work, so I'm fine with that for now.

Finally,  in _start_dependancies:

```python
for dependency in self._container_dependencies.get(container, []):
```

becomes

```
for dependency in container.depends_on:
```

That's exactly the same logic, but dependencies are shipped directly in the container instance, rather than in a separated dictionnary, which will make the code easier to understand an maintain.

Reply:
At that point I'm thinking of even moving the parallel start logic to `TestedContainer` itself as well, WDYT?

Reply:
It wouldn't work for multiple root containers, but at the end of the day the goal is to start weblog, so weblog can start all of its dependencies itself at that point.

Reply:
Actually it would work even then since at that point `start` would return its thread that could then be joined.

Reply:
> At that point I'm thinking of even moving the parallel start logic to TestedContainer itself as well, WDYT?

but 

> It wouldn't work for multiple root containers

The code that works on all situations is the one to go. If one day we have two root containers, we'll need to revert.

Reply:
The original comment has now been addressed, and the code has been updated to reflect our latest discussions.